ZEP-XXX-journal-record-format.md

```diff
@@ -0,0 +1,239 @@
+
+A replicated record contains following fields:
+1. term (long)
+2. index (long)
+3. asqn (asqn)
+4. checksum (long)
+5. serializedRaftRecord (byte[])
+
+Here `serializedRaftRecord` is the sbe serialized bytes of RaftRecord with the following schema.
+```
++ --------------------------------------------------
+| term<uint64> | entry type | entry<SBE(RaftEntry)> |
++ --------------------------------------------------
+```
+
+The raft thread that receives this record uses the term and index to verify the preconditions before writing it to the journal. It should then construct a journal record using `index`, `asqn`, `checksum` and `serializedRaftRecord` which can be appended to the journal. Journal is expected to verify if the index and checksum is correct.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+### Alternatives considered
+
+1. RaftRecord does not have the field entry type, instead we can determine the type of the entry from the sbe schema.
+With this approach, there will be an empty InitialEntry with only sbe headers and no fields. This approach is also fine.
+However, having an entry type make its explicit. It might be also useful when adding new types or changing the existing types.
+2. Instead of frame, we use the checksum field to mark a record is invalid. If the checksum is 0, we assumes that it is EOF. While this works for detecting corruption, we decided to go with the frame because of its additional benefits.
+
+
+### Why do we need a Frame?
+
+SBE can handle versions. We can add or remove fields (with some limitations).
+However adding a frame before the record has the following benefits:
+1. Marking valid records and marking end of file.
+2. Encoding the version number to the frame allow us to change the serialization format from SBE to another without breaking compatability.
+3. By encoding version number to the frame, we are more flexible in changing the journal record format.
+
+### Why did we add index and asqn to the record?
+The index and asqn are concepts defined in the journal. They are part of a valid journal record.
+
+Previously, the index was not written together with the record. Instead it was calculated.
+But this is error prone.
+There is no way to detect if the log is corrupted by having an entry at the wrong index.
+It is safer to persist the index with the record.
+
+It is the responsibility of the journal to keep the mappings between the index and the asqn.
+Hence, it should be possible for the journal to determine the asqn without the need for reading the raft entry.
+
+### How should versioning work in future?
+SBE supports versioning.
+As long as the adding or removing fields are done according to what SBE supports, it is possible to handle multiple versions.
+Multiple versions should be handled in the journal writer or reader.
+
+For any change in the layout that is not supported by SBE, we can make use of the frame by encoding the version number.
```

Comment:
Would it not work to simply introduce a new SBE template? Then we need the code to handle, say, for application entry template 5 and 7? And they can just be identified by their template ID? Am I missing something? I imagine the version number is mostly to let us switch away from SBE?

Reply:
It is possible, I think. We can define new templates if there are major changes to the schema. Then the serializer should be able to handle it.